readers/hdfs/src/main/java/com/criteo/hadoop/garmadon/hdfs/offset/HdfsOffsetComputer.java

```diff
@@ -45,51 +48,87 @@ public HdfsOffsetComputer(FileSystem fs, Path basePath, @Nullable String kafkaCl
-                        result.merge(partitionId, offset, Long::max);
+        final Map<Integer, Map<String, FinalEventPartitionFile>> resultFile = new HashMap<>();
+
+        LocalDateTime today = LocalDateTime.now();
+        for (int i = 0; i < backlogDays; ++i) {
+            LocalDateTime day = today.minusDays(i);
+            String listedDay = day.format(DateTimeFormatter.ISO_DATE);
+
+            String dirsPattern = String.format("%s/*", computeDirName(day));
+            FileStatus[] fileStatuses = fs.globStatus(new Path(basePath, dirsPattern));
+
+            for (FileStatus status : fileStatuses) {
+                String fileName = status.getPath().getName();
+                Matcher matcher = offsetFilePatternGenerator.matcher(fileName);
+                if (matcher.matches()) {
+                    try {
+                        int partitionId = Integer.parseInt(matcher.group("partitionId"));
+                        Long index = Long.parseLong(matcher.group("index"));
+                        if (partitionIdsSet.contains(partitionId) &&
+                                (!resultFile.containsKey(partitionId) || !resultFile.get(partitionId).containsKey(listedDay) ||
+                                        index > resultFile.get(partitionId).get(listedDay).getIndex())) {
+                            HashMap<String, FinalEventPartitionFile> dateFinalEventPartitionFile =
+                                    (HashMap<String, FinalEventPartitionFile>) resultFile.computeIfAbsent(partitionId, s -> new HashMap<>());
+                            dateFinalEventPartitionFile.put(listedDay,
+                                    new FinalEventPartitionFile(index, status.getPath()));
+                            resultFile.put(partitionId, dateFinalEventPartitionFile);
+                        }
+                    } catch (NumberFormatException e) {
+                        LOGGER.info("Couldn't deviate a valid offset from '{}'", fileName);
                     }
-                } catch (NumberFormatException e) {
-                    LOGGER.info("Couldn't deviate a valid offset from '{}'", fileName);
                 }
             }
         }
 
+        // Get last offset
+        for (int partitionId : resultFile.keySet()) {
+            result.put(partitionId, getMaxOffset(resultFile.get(partitionId)));
+        }
+
         return result;
     }
 
-    private String computeDirNamesPattern(LocalDateTime today, int backlogDays) {
-        List<String> dirnames = new ArrayList<>(backlogDays);
-
-        for (int i = 0; i < backlogDays; ++i) {
-            LocalDateTime day = today.minusDays(i);
-
-            dirnames.add(computeDirName(day));
+    protected Long getMaxOffset(Map<String, FinalEventPartitionFile> dateFinalEventPartitionFile) throws IOException {
+        List<Long> maxOffsets = new ArrayList<>();
+        for (FinalEventPartitionFile finalEventPartitionFile : dateFinalEventPartitionFile.values()) {
```

Comment:
you could also stream dateFinalEventPartitionFile.values() instead of iterating. Streams can be parallelised, is it something we could benefit since we might have many files to read ?

Reply:
We could have 2 files to read in // with current setup so yes lets use stream